src/pytest_benchmark/elasticsearch_storage.py

```diff
@@ -0,0 +1,277 @@
+import datetime
+
+
+class ElasticsearchStorage(object):
+    def __init__(self, elasticsearch_hosts, elasticsearch_index, elasticsearch_doctype, logger,
+                 default_machine_id=None):
+        try:
+            import elasticsearch
+        except ImportError as exc:
+            raise ImportError(exc.args, "Please install elasticsearch or pytest-benchmark[elasticsearch]")
+        self._elasticsearch_hosts = elasticsearch_hosts
+        self._elasticsearch_index = elasticsearch_index
+        self._elasticsearch_doctype = elasticsearch_doctype
+        self._elasticsearch = elasticsearch.Elasticsearch(self._elasticsearch_hosts)
+        self.default_machine_id = default_machine_id
+        self.logger = logger
+        self._cache = {}
+        self._create_index()
+
+    def __str__(self):
+        return str(self._elasticsearch_hosts)
+
+    @property
+    def location(self):
+        return str(self._elasticsearch_hosts)
+
+    def query(self, project):
+        """
+        Returns sorted records names (ids) that corresponds with project.
+        """
+        return [commit_and_time for commit_and_time, _ in self.load(project)]
+
+    def load(self, project, id_prefix=None):
+        """
+        Yield key and content of records that corresponds with project name.
+        """
+        r = self._search(project, id_prefix)
+        groupped_data = self._group_by_commit_and_time(r["hits"]["hits"])
+        result = [(key, value) for key, value in groupped_data.items()]
+        result.sort(key=lambda x: datetime.datetime.strptime(x[1]["datetime"], "%Y-%m-%dT%H:%M:%S.%f"))
+        for key, data in result:
+            yield key, data
+
+    def _search(self, project, id_prefix=None):
+        body = {
+            "size": 1000,
+            "sort": [
+                {
+                    "datetime": {
+                        "order": "desc"
+                    }
+                }
+            ],
+            "query": {
+                "bool": {
+                    "filter": {
+                        "term": {
+                            "commit_info.project": project
+                        }
+                    }
+                }
+            }
+        }
+        if id_prefix:
+            body["query"]["bool"]["must"] = {
+                "prefix": {
+                    "_id": id_prefix
+                }
+            }
+
+        return self._elasticsearch.search(index=self._elasticsearch_index,
+                                          doc_type=self._elasticsearch_doctype,
+                                          body=body)
+
+    @staticmethod
+    def _benchmark_from_es_record(source_es_record):
+        result = {}
```

Comment:
You can see how single record looks in es in the bottom. I had to split benchmarks of each run to single record and therefore put common information (e.g. `machine_info`, `datetime`) to each record. Function `_benchmark_from_es_record` selects only keys that belongs to single benchmark, while `_run_info_from_es_record` selects data that are common among these benchmarks.

``` {
            "_index": "benchmark",
            "_type": "benchmark",
            "_id": "ukuran_6eac319724dbd9beebf21aafd52d00ece31f61a4_20160912_080154_tests/test_performance.py::test_pedantic",
            "_score": null,
            "_source": {
               "version": "3.0.0",
               "options": {
                  "timer": "perf_counter",
                  "warmup": false,
                  "disable_gc": false,
                  "max_time": 1,
                  "min_rounds": 5,
                  "min_time": 0.000005
               },
               "commit_info": {
                  "project": "ukuran",
                  "id": "6eac319724dbd9beebf21aafd52d00ece31f61a4",
                  "dirty": false
               },
               "stats": {
                  "stddev": 0.0000033206052045174887,
                  "q1": 0.005085450800015679,
                  "stddev_outliers": 2,
                  "q3": 0.005090232499969716,
                  "rounds": 10,
                  "median": 0.005088654999963182,
                  "max": 0.0050920920999487865,
                  "iqr": 0.000004781699954037148,
                  "outliers": "2;0",
                  "min": 0.005080863300008787,
                  "iqr_outliers": 0,
                  "ld15iqr": 0.005080863300008787,
                  "mean": 0.005087793719976617,
                  "iterations": 10,
                  "hd15iqr": 0.0050920920999487865
               },
               "machine_info": {
                  "node": "ubuntu",
                  "processor": "x86_64",
                  "machine": "x86_64",
                  "system": "Linux",
                  "release": "4.4.0-36-generic",
                  "python_implementation": "CPython",
                  "python_build": [
                     "default",
                     "Jul  5 2016 12:43:10"
                  ],
                  "python_implementation_version": "3.5.2",
                  "python_compiler": "GCC 5.4.0 20160609",
                  "python_version": "3.5.2"
               },
               "group": null,
               "fullname": "tests/test_performance.py::test_pedantic",
               "name": "test_pedantic",
               "datetime": "2016-09-12T08:01:57.865021",
               "param": null,
               "params": null
            },
            "sort": [
               1473667317865
            ]
         }```